dataflow/logging: Avoid count in log analysis, perf improvements

antiguru opened the pull request:
Signed-off-by: Moritz Hoffmann <mh@materialize.com>

### Motivation

This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/2208

### Description

The logging introspection implementation maintains counted events instead of relying on arrangements maintaining the counts. This means the logging infrastructure causes more work to maintain, even when not being used.

With this change, the effort to evaluate the underlying tables is lifted to the user. By adding transparent views, existing queries should function without changes.

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

antiguru commented:
This currently fails for the following query. Mz will OOM.
```
SELECT
    id, worker, count(*) AS elapsed_ns
FROM
    mz_catalog.mz_scheduling_elapsed_internal
GROUP BY
    id, worker;
```

antiguru commented:
I'm not very fond of the `_internal` tables, which need to live somewhere but might confuse users and crash Mz when queried without aggregation. Do we have any way to hide logs or move them to the danger zone?

antiguru commented:
> This currently fails for the following query. Mz will OOM.
> 
> ```
> SELECT
>     id, worker, count(*) AS elapsed_ns
> FROM
>     mz_catalog.mz_scheduling_elapsed_internal
> GROUP BY
>     id, worker;
> ```

The problem was that the log still had key columns defined, which confuses Mz. Removing the keys fixed the OOM.

antiguru commented:
I'll add similar changes for Differential and Mz logging tomorrow. Reviews are not urgent!

antiguru commented on src/coord/src/catalog/builtin.rs:
```diff
@@ -1419,24 +1520,25 @@ lazy_static! {
-            Builtin::Log(&MZ_SCHEDULING_HISTOGRAM),
-            Builtin::Log(&MZ_SCHEDULING_PARKS),
-            Builtin::Log(&MZ_ARRANGEMENT_SIZES),
-            Builtin::Log(&MZ_ARRANGEMENT_SHARING),
+            Builtin::Log(&MZ_DATAFLOW_OPERATOR_REACHABILITY_INTERNAL),
+            Builtin::Log(&MZ_KAFKA_BROKER_RTT),
+            Builtin::Log(&MZ_KAFKA_CONSUMER_PARTITIONS),
             Builtin::Log(&MZ_MATERIALIZATIONS),
             Builtin::Log(&MZ_MATERIALIZATION_DEPENDENCIES),
-            Builtin::Log(&MZ_WORKER_MATERIALIZATION_FRONTIERS),
+            Builtin::Log(&MZ_MESSAGE_COUNTS_RECEIVED_INTERNAL),
+            Builtin::Log(&MZ_MESSAGE_COUNTS_SENT_INTERNAL),
             Builtin::Log(&MZ_PEEK_ACTIVE),
             Builtin::Log(&MZ_PEEK_DURATIONS),
+            Builtin::Log(&MZ_SCHEDULING_ELAPSED_INTERNAL),
+            Builtin::Log(&MZ_SCHEDULING_HISTOGRAM_INTERNAL),
+            Builtin::Log(&MZ_SCHEDULING_PARKS_INTERNAL),
             Builtin::Log(&MZ_SOURCE_INFO),
-            Builtin::Log(&MZ_MESSAGE_COUNTS),
-            Builtin::Log(&MZ_KAFKA_CONSUMER_PARTITIONS),
-            Builtin::Log(&MZ_KAFKA_BROKER_RTT),
+            Builtin::Log(&MZ_WORKER_MATERIALIZATION_FRONTIERS),
```
Sorted Logs and Views below. Can revert if wanted.

uce reviewed (approved):
LGTM.

* Updated built-in sources and views ✅ 
* Logging updates ✅ (the `logging/timely.rs` changes were the toughest to grok for me)
* Chaining ✅ 
* Avoid string allocs ✅
* nit: review might have been a bit simpler for me if we had separate commits for the changes (no biggie)
* nit: It might be insightful for others to share the numbers you mentioned offline

antiguru commented:
Noting down some performance data. I'm running Materialize with the following configuration: `--workers 16 --introspection-frequency 1s --debug-introspection`. On a clean `mzdata` directory, I run the following script:

```
DROP DATABASE IF EXISTS MATERIALIZE;
CREATE DATABASE MATERIALIZE;
create materialized view test2 as select generate_series as x from generate_series(1, 10);
select count(*) from test2;
```

The select statement is repeated 10k times. I measure the time it takes for `pgsql` to process all queries and observe the Rss.

Materialize main: 30s, 2.9GiB Rss

Materialize with patch: 29s, 2.9GiB Rss

So, from this perspective the change is not clearly an improvement.

antiguru commented:
I pulled the changes to enable worker-local pre-aggregation into the PR because I think only this way it makes sense to merge this change.

Materialize with additional changes: 24s, 1GiB Rss

antiguru commented:
The latest iteration of this PR changes the semantics of some log arrangements. While in the past they would potentially know about all changes, even short lived ones, this is not the case anymore. Logs that needed derived data to retract metrics would keep a delayed log of events to remove data later, which would also mean that all changes would be observable from the outside. Other logs would pass all changes to the arrangements, but since both creation and deletion could occur during the same introspection interval, the data would cancel out each other and thus would never get arranged.

The worker-local pre-aggregation immediately removes events that cancel out each other. For example, if within the introspection interval an operator is created and dropped, its logs will sum up to zero and will be removed by the pre-aggregation. As a result, more data would be hidden from the user\, but likely with a negligible change of semantics.

frankmcsherry commented on src/dataflow/src/logging/mod.rs:
```diff
@@ -109,3 +121,89 @@ where
+        self.buffer.push(data);
+        if self.buffer.len() == self.buffer.capacity() {
+            // Consolidate while the consolidation frees at least half the buffer
+            if self.clean < self.buffer.capacity() / 2 {
+                consolidate_updates(&mut self.buffer);
+                self.clean = self.buffer.len();
+            } else {
+                // Consolidate once more and flush
+                consolidate_updates(&mut self.buffer);
+                self.flush();
+            }
```
I think this logic may be slightly off, in that it seems possible that a consolidation happens, the buffer remains full, and is not flushed. A different way is, since you always call `consolidate_updates(&mut self.buffer);` is to just call that first, look at the results, and either ship or do not based on whether the buffer is at least half full.

antiguru replied:
Thanks for pointing this out! I first implemented a scheme where the buffer[clean..] would be consolidated up until the buffer was full (or above a threshold), and then it would try to consolidate everything. But it's kind of hard to analyze its performance, so I reverted to a simpler approach. I updated the PR to remove the remains of `clean`.

uce reviewed (approved):
LGTM.

* From one of your commit messages:
  > The downside it that the implementation is less elegant and
  requires potentially more care when maintaining.
  
  I agree with this and I think it's ok since we don't touch this code frequently and the reduction in memory consumption seems to be worth it.

* What do you think about adding a release note about the cases where this slightly changes what users can observe? 

* I'd be curious to see a comparison of the release load tests comparing main, this patch, and logging turned off.

uce commented on src/coord/src/catalog/builtin.rs:
```diff
@@ -1361,6 +1401,149 @@ pub const PG_SETTINGS: BuiltinView = BuiltinView {
+    needs_logs: true,
+};
+
+pub const MZ_ARRANGEMENT_SHARING: BuiltinView = BuiltinView {
+    name: "mz_arrangement_sharing",
+    schema: MZ_CATALOG_SCHEMA,
+    sql: "CREATE VIEW mz_arrangement_sharing AS
+SELECT
+    operator,
+    worker,
+    pg_catalog.count(*) AS count
+FROM mz_catalog.mz_arrangement_sharing_internal
+GROUP BY operator, worker",
+    id: GlobalId::System(5033),
+    needs_logs: true,
+};
```
I hadn't caught that count view was missing in the earlier commit. Sorry!

What do you think about extending `logging.td` with queries for these?

uce replied:
Just to be clear: the current commit LGTM. I'm mainly asking for the tests 😄

I think we had a related case in https://github.com/MaterializeInc/materialize/pull/6445#pullrequestreview-634867678.

antiguru replied:
Added a test to the PR.

antiguru commented on src/dataflow/src/logging/mod.rs:
```diff
@@ -109,3 +121,82 @@ where
+/// The buffer is filled with updates until it reaches its capacity. At this point, the updates are
+/// consolidated to free up space. This process repeats until the consolidation recovered less than
+/// half of the buffer's capacity, at which point the buffer will be filled, once more consolidated
+/// and shipped.
```
Updated and expanded the comment. Thanks for noticing it!

antiguru commented:
I added some release notes and addressed the remaining comments. A big missing item is the integration with load tests, which I think should be captured as a separate issue as it's outside the scope of this effort.

uce commented on test/testdrive/logging.td:
```diff
@@ -55,3 +55,142 @@ cannot drop schema mz_catalog because it is required by the database system
 ! DROP VIEW mz_peek_durations
 cannot drop item mz_catalog.mz_peek_durations because it is required by the database system
+
+? EXPLAIN TYPED PLAN FOR SELECT operator, worker, count FROM mz_arrangement_sharing
```
LGTM. I had asked @philip-stoev about how to check the existence of these columns without requiring to provide the expected result and he had `COUNT(*) >= 0` trick up his sleeve.

```
# Verify that all listed columns are available. COUNT(*) >= 0 allows us to avoid
# providing the exact result to testdrive.
> SELECT COUNT(*) >= 0 FROM (SELECT operator, worker, count FROM mz_arrangement_sharing);
true
```

antiguru replied:
I replaced the test with a lookup of `mz_views` and `mz_columns`. Thanks to @philip-stoev for pointing this out!

uce replied:
That seems indeed better!

uce commented on doc/user/content/release-notes.md:
```diff
@@ -48,6 +48,13 @@ Wrap your release notes at the 80 character mark.
 
 {{% version-header v0.9.5 %}}
 
+- Improve the performance of the log introspection, which should result in
+  lower latency spikes increased throughput while issuing many small queries,
+  and lower memory consumption. Additionally, with this change, the content
+  of the views are consistent at the introspection interval boundaries. (In the
+  past, some logs would reveal more details about ephemeral dataflows and
+  operators.)
```
```suggestion
- Improve the performance of [built-in sources and views](/cli/#introspection-sources)
  in the [system catalog](/sql/system-catalog), which should result in lower
  latency spikes and increased throughput when issuing many small queries,
  and reduce overall memory consumption. Additionally, the content of the views
  is now consistent at the introspection interval boundaries. Prior to this
  release, some views would reveal more details about ephemeral dataflows and
  operators.
```

I mainly wanted to respect

> Link to at least one page where users can learn more about either the change or the area which the change was made.

from the release notes guide (a comment at the top of the file) but then ended up changing some wording. Please feel free to ignore.

antiguru replied:
Thanks for the suggestions, accepted!

uce reviewed (approved):
LGTM.

Thanks for adding the test. I left two more optional comments that should not block merging (one about a more concise way to test the columns and one about the release notes).

antiguru commented:
Thanks a bunch for the reviews!
